### PR TITLE
Simple cve-search integration PoC

### PR DESCRIPTION
**Pull request by StCyr**
## Status
**IN DEVELOPMENT**

## Description
CVE-Search integration PoC

This PoC can be tested by installing a cve-search instance on the host running OCSReport.

The cve-search instance must be patched to support CORS (see https://github.com/cve-search/cve-search/pull/300), and the HTTP server of OCSreport must set the following header: `Content-Security-Policy "default-src 'self' 'unsafe-inline' http://127.0.0.1:5000;`"

## Related Issues
https://github.com/OCSInventory-NG/OCSInventory-ocsreports/issues/593

## Todos
- [X] Tests
- [X] Documentation

## Test environment
OCSI 2.5 on Apache (with Content-Security_policy support)
cve-search (https://github.com/cve-search/cve-search) as of commit 7b23434c7b06df6160089e64276fc92b2242c2fb (patched to support CORS)

## Impacted Areas in Application
List general components of the application that this PR will affect:

* DataTables
* Software tables

### CODE REVIEW

**Comment by StCyr** [comment]
just in case, the icon was found here: https://cdn4.iconfinder.com/data/icons/social-messaging-productivity-4/128/bug-128.png

**Comment by gillesdubois** [comment]
Hi @StCyr,

In the first place, your contribution is amazing ! That feature will be a big plus when implemented to OCS :D 

I just have one problem with your implementation, the CVE search URL is hardcoded. We need to find a solution to make it dynamic.

Your contribution is really nice and can evolve in the future.
We could create a dashboard with a summary of all vulnerables softwares in our assets. 

Concerning your question, I don't think it's a good idea to use tooltip using datatable. By the way, can you send me what code did you use for your first try ?

If you need help don't hesitate to ask.

Regards,
Gilles Dubois.

**Comment by gillesdubois** [comment]
@StCyr : Take care, your last commit removed changes that are validated in the current master branch.

Concerning, the tooltip i will come back to you asap

Regards.

File: require/function_table_html.php [hunk 1]
```diff
@@ -536,20 +536,25 @@ function CpeNormalizeVendor(vendor) {
                 return vendor;
             };
             // get CVE info from a CVE-Search instance for a given software
-            function getCVE(row) {
```
Comment (charleneauger) [inline require/function_table_html.php]: Hi @StCyr ,
The deletion of this function is wanted or is it just a mistake?

Regards,
Charlene Auger

File: plugins/main_sections/ms_config/ms_config.php [hunk 2]
```diff
@@ -82,11 +82,15 @@
   $def_onglets['FILTER'] = $l->g(735); //Filter
   $def_onglets['WEBSERVICES'] = $l->g(760); //Webservice
   $def_onglets['CNX'] = $l->g(1108); //connexion
+  $def_onglets['VULN'] = $l->g(1450); //cve-search integration
```
Comment (charleneauger) [inline plugins/main_sections/ms_config/ms_config.php]: Hi @StCyr ,

Your translation number is wrong...

Regards,
Charlene

File: require/function_config_generale.php [hunk 3]
```diff
@@ -871,6 +871,21 @@ function pagewebservice() {
...
+
+    // Which lines do we need?
+    $champs = array('VULN_CVESEARCH_ENABLE' => 'VULN_CVESEARCH_ENABLE',
+        'VULN_CVESEARCH_HOST' => 'VULN_CVESEARCH_HOST');
+
+    // Get configuration values from DB
+    $values = look_config_default_values($champs);
+
+    // Display configuration items
+    ligne('VULN_CVESEARCH_ENABLE', $l->g(1451), 'radio', array(1 => 'ON', 0 => 'OFF', 'VALUE' => $values['ivalue']['VULN_CVESEARCH_ENABLE']));
```
Comment (charleneauger) [inline require/function_config_generale.php]: Hi @StCyr ,

Your translation number is wrong...

Regards,
Charlene

File: require/function_config_generale.php [hunk 4]
```diff
@@ -871,6 +871,21 @@ function pagewebservice() {
...
+    ligne('VULN_CVESEARCH_HOST', $l->g(1452), 'input', array('VALUE' => $values['tvalue']['VULN_CVESEARCH_HOST'], 'SIZE' => "30%", 'MAXLENGTH' => 200));
```
Comment (charleneauger) [inline require/function_config_generale.php]: Hi @StCyr ,

Your translation number is wrong...

Regards,
Charlene

**Comment by StCyr** [comment]
Hi @charleneauger,

thanks for your review. I'm now fixing my mistakes.

Cyrille

**Comment by StCyr** [comment]
btw, it looks like the bug icon I'm using is not free. I propose to use this one: https://www.iconsdb.com/custom-color/bug-3-icon.html

**Comment by StCyr** [comment]
also, I'm setting-up a public cve-search instance that you'll be able to use to validate my PR. I'll let you know the url

**Comment by charleneauger** [comment]
Hi @StCyr ,

No need to set a cve-search public instance, I installed it on a server and I could test your PR. It's really good for the moment ! 

Regards,
Charlene

**Comment by StCyr** [comment]
Hi,

I've updated my work quite a bit: getCVE is now a recursive function that tries up to 3 naming scheme when querying the cve-search instance.

I'm not happy with the icons' color (the yellow is awfull). But, I'm not happy with the color code neither. 

For the moment the color code is:
* red: one or several cve has/have been found with the exact software version found
* orange: one or several cve has/have been found without the exact software version found
* yellow: no cve found
* grey: queries pending

I think it should be:
* red: one or several CVE applicable to the software version or above has/have been found 
* orange: no CVE found (I would put it in orange because it may be that we didn't found any CVE because the software name reported by OCSI is too different from its CPE equivalent)
* green: one or several CVEapplicable to a lower version has/have been found (I would put it green because the fact that we found at least one CVE is a good indicator that the CPE queried is correct)
* grey: queries pending

Cyrille

**Comment by StCyr** [comment]
Hi @charleneauger, @gillesdubois,

I think my work is 95%+ done. Could you give me an updated feedback?

Thanks

Cyrille

File: require/function_table_html.php [hunk 5]
```diff
@@ -544,6 +670,40 @@ function checkall()
...
+                                    cveAjaxRequests[i].abort();
+                                }
+                                cveAjaxRequests = [];
+                            }
+                            cveSearchUrl = "<?php $h=look_config_default_values(array('VULN_CVESEARCH_HOST'));echo $h['tvalue']['VULN_CVESEARCH_HOST'] ?>";
+                            // For data rows only
+                            $(table_id).find(".odd,.even").each(function(){
+                                // Get software name
+                                var soft = $(this).find(".NAME,.NAME sorting_1").text();
+                                soft = cpeNormalizeName(soft);
```
[the remark this answers is not in the document]
Comment (StCyr) [reply require/function_table_html.php]: fixed

File: require/function_table_html.php [hunk 6]
```diff
@@ -544,6 +670,40 @@ function checkall()
...
                     }
+                    // For tables listing softwares, try to get CVE info
+                    if (table_id == 'table#affich_soft') {
+                        // Only if cve-search integration is enabled
+                        var cveEnabled = "<?php $h=look_config_default_values(array('VULN_CVESEARCH_ENABLE'));echo $h['ivalue']['VULN_CVESEARCH_ENABLE'] ?>";
+                        if (cveEnabled==1) {
+                            // Abort any pending ajax requests
+                            if (cveAjaxRequests.length>0) {
+                                for (i=0;i<cveAjaxRequests.length;i++){
+                                    cveAjaxRequests[i].abort();
```
Comment (charleneauger) [inline require/function_table_html.php]: Hi @StCyr ,

When I change "show entries" , this error appears in my console : `Uncaught TypeError: cveAjaxRequests[i].abort is not a function`

Regards,
Charlene

File: require/function_table_html.php [hunk 7]
```diff
@@ -516,8 +518,132 @@ function checkall()
...
+                        var nameColumn = this.find(".NAME,.NAME sorting_1");
+                        // Get back software infos (we cannot get them from the table because they might have changed during the recursion)
+                        var softInfo=getSoftwareInfoFromJqxhr(cveAjaxRequests,jqxhr);
+                        var soft=softInfo[0];
+                        var vendor=softInfo[1];
+                        var version=softInfo[2];
+                        // Did we found any CVE?
+                        if (data.length>0) {
+                            // CVE found, let's first consider none match this software version or above
+                            var thisCveUrl = cveSearchUrl + '/search/' + vendor + "/" + soft;
```
Comment (charleneauger) [inline require/function_table_html.php]: Hi @StCyr ,

I think you forgot `/api ` in the url :)

Regards,
Charlene

Comment (StCyr) [reply require/function_table_html.php]: No, without the `/api` part, the URL points to a more human-readable page.

File: require/function_table_html.php (hunk 6 above)
Comment (StCyr) [reply require/function_table_html.php]: fixed

**Comment by StCyr** [comment]
Hello,

I think I solved all the issues, except 2 or 3 that I will list at the bottom of this comment.

Though, I'm still not very happy with the bug icons' color code. I wonder if the following wouldn't be better:

- grey, fluttering: query pending
- grey, fix: no CVE found (not using the green color because it may be that we didn't found any CVE because the software name reported by OCSI is too different from its CPE equivalent)
- red: one or several CVE applicable to the software version or above has/have been found
- orange: one or several CVEapplicable to a lower version has/have been found (I don't like using the green color because I'm afraid the version comparaison function might be wrong, while the green color would give the user a false sense of safety)
- grey, crossed: error when trying to query the cfe-search server.

I wonder also about implementing some color scale, from green to red via orange. Example:

- green (#33cc33): at most 5 CVE found, none matching this version or above
- clear green (#ccff99): at most 50 CVE found, none matching this version or above
- orange (#ffcc66): at least 50 CVE found, none matching this version or above
- light read (#ff6666): at most 5 CVE found  matching this version or above
- red (#ff0000): at least 5 CVE found matching this version or above

Maybe, this color scale should also take the CVE severity into account:

- green (#33cc33): no CVE found matching this version or above. Total of severity is less than 10
- clear green (#ccff99): no CVE found matching this version or above. Total of severity is between 10 and 50
- orange (#ffcc66): no CVE found matching this version or above. Total of severity is more than 50
- light read (#ff6666): CVE found  matching this version or above. Total of severity is less than 10
- red (#ff0000): CVE found  matching this version or above. Total of severity is more than 10

The numbers are obviously indicative only. And, maybe, I should foresee to make them configurable.

Now, for the 2-3 remaining issues that I would like to fix:

1. When clicking on a bug icon, the new web page should open in another tab.
2. I would like to implement a special case for identifying the software "Java Runtime Environment" which is reported as something like "Java 8 Update 181 (64-bit)" ("Java 8 Update" after CPE normalization) while it is identified as "oracle:jre" in the CPE dictionary.
3. I still think it would be good if I could implement tooltips on the bug icons, showing, for example, the amount of CVE matching 

That's it.

Bye

**Comment by gillesdubois** [comment]
Hi @StCyr,

In the first place thx for this amazing contribution.
However, do you think you can rebase to the master branch ?
After some check i saw that some of the code from the master is erased in your commits.

For example in config page, you remove the 'INVENTORY_SAAS_ENABLED' which is in the master.

Also, we won't be able to merge this until we have made the documentation for this si it may take some time. We will create a PR in https://github.com/OCSInventory-NG/Wiki 

If you already have some documentation about what you've done don't hesitate to send us a mail to : contact@ocsinventory-ng.org

Regards,
Gilles Dubois.

**Comment by StCyr** [comment]
Hello,

Sure, i'll rebase my work.

Sorry for the unintentional erasings; It's because of my development
workflow: I first develop in my live server, then copy my changes manually
to my local git repo when my changes look good. This is obsviously error
prone :-(

I don't have any documentation yet but I don't mind writing it if you want.

Br,

Cyrille




Le mer. 14 nov. 2018 à 10:40, Gilles Dubois <notifications@github.com> a
écrit :

**Comment by StCyr** [comment]
Hello @gillesdubois 

I've rebased my PR. 

I've identified 1 line that I'd mistakenly removed from file function_config_generale.php

**Comment by StCyr** [comment]
Hello,

I just wanted to let you know that I wrote a similar integration to make sure the software versions installed are the latest one:

image

Color code meaning:

- Green : Latest version
- Red: Not latest version
- Orange: not found

I find it more useful than this CVE-Search integration.

**Comment by gillesdubois** [comment]
Hi,

Thx for your contribution, we re-adapted the code to implement a more generic way of dealing with the CVE search. The pull request has been merged.

Regards,
Gilles Dubois.

**Pull request closed without merging**